arm: soc: nordic nRF52: Add MPU support

mike-scott opened the pull request:
We now have generic ARM M4 MPU support added to Zephyr.
Let's enable it for use with Nordic nRF52 chips.

Memory Layout was generated from Section 8.3 "Memory
Map" of nRF52 Product Specifications (for both nRF52832
and nRF52840):
0x00000000: Flash
0x10000000: Factory Information Config Registers
0x10001000: User Information Config Registers
0x20000000: SRAM
0x40000000: APB Peripherals
0x50000000: AHB Peripherals
0xE0000000: ARM M4 Private Peripheral Registers

NOT Configured:
0x60000000: External RAM
0x80000000: External RAM
0xA0000000: External Device
0xC0000000: External Device

NOTE: More work will be needed for future Nordic MWU (Memory
Watching Unit) support.

Change-Id: I6d2629bb70abc8c85dc81fdf7e03e7be01478ff6
Signed-off-by: Michael Scott <michael.scott@linaro.org>

Commit:
arm: soc: nordic nRF52: Add MPU support

We now have generic ARM M4 MPU support added to Zephyr.
Let's enable it for use with Nordic nRF52 chips.

Memory Layout was generated from Section 8.3 "Memory
Map" of nRF52 Product Specifications (for both nRF52832
and nRF52840):
0x00000000: Flash
0x10000000: Factory Information Config Registers
0x10001000: User Information Config Registers
0x20000000: SRAM
0x40000000: APB Peripherals
0x50000000: AHB Peripherals
0xE0000000: ARM M4 Private Peripheral Registers

NOT Configured:
0x60000000: External RAM
0x80000000: External RAM
0xA0000000: External Device
0xC0000000: External Device

NOTE: More work will be needed for future Nordic MWU (Memory
Watching Unit) support.

Signed-off-by: Michael Scott <michael.scott@linaro.org>

mike-scott commented on arch/arm/soc/nordic_nrf5/nrf52/arm_mpu_regions.c:
```diff
@@ -0,0 +1,47 @@
+#include "arm_mpu_mem_cfg.h"
+
+#define FUCR_BASE	0x10000000
+#define APB_PERIPH_BASE	0x40000000
```
This looks like there's no tab here.. but in the code there is.  Odd.  (Same below)

cvinayak commented on arch/arm/soc/nordic_nrf5/nrf52/arm_mpu_regions.c:
```diff
@@ -0,0 +1,47 @@
+
+#include "arm_mpu_mem_cfg.h"
+
+#define XICR_BASE	0x10000000
```
Use spaces instead of tabs

galak replied:
where? tabs are normal default

carlescufi replied:
@galak he means between XICR_BASE and 0x10000000, that's what we do in the bluetooth subsystem

cvinayak commented on arch/arm/soc/nordic_nrf5/nrf52/Kconfig.soc:
```diff
@@ -26,3 +26,11 @@ config SOC_NRF52840_QIAA
+	select ARM_MPU
+	default n
+	help
+	  Enable MPU
```
Enable MPU support on Nordic Semiconductor nRF52x series ICs.

mike-scott replied:
Nice catch.  Fixing w/ your text.

cvinayak commented on arch/arm/soc/nordic_nrf5/nrf52/Makefile:
```diff
@@ -10,6 +10,7 @@ soc-cflags += -DNRF52840_XXAA
 endif
 
 obj-y += soc.o
+obj-$(CONFIG_NRF52_ARM_MPU_ENABLE) += arm_mpu_regions.o
```
since the file is under arch/arm path, just mpu_regions.c as filename?

mike-scott replied:
Both the ARM Beetle and STM32 implementation used this naming structure.  You have a good point however.

carlescufi replied:
This might be because it refers to ARM's standard MPU, as opposed to a custom MPU implemented by the IC vendor

fvincenzo replied:
mpu_regions is the configuration file for a specific mpu, to easily identify the vendor, the format of the configuration file name currently is <vendor>_mpu_regions and the driver filename is <vendor>_mpu.

cvinayak commented on arch/arm/soc/nordic_nrf5/nrf52/arm_mpu_mem_cfg.h:
```diff
@@ -0,0 +1,42 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef _ARM_MPU_MEM_CFG_H_
```
Same here just mpu_mem_cfg.h ?

cvinayak commented on arch/arm/soc/nordic_nrf5/nrf52/Kconfig.soc:
```diff
@@ -26,3 +26,11 @@ config SOC_NRF52840_QIAA
 
 endchoice
+
+config NRF52_ARM_MPU_ENABLE
```
ARM_MPU_NRF52X

mike-scott replied:
The STM32 implementation uses STM32_ARM_MPU_ENABLE, so I also used that pattern in this patch.  However, the ARM Beetler uses "ARM_MPU_ENABLE" (w/o SoC at all) so there is no standard, I'm open to using your format.

mike-scott replied:
Clarification: Did you mean ARM_MPU_NRF52X_ENABLE?

cvinayak replied:
_ENABLE is unnecessary, and makes the define long. The feature is enabled if it is defined to 'y'; 

CONFIG_ARM_MPU_NRF52X=y

This is sufficient to convery, ARM arch MPU in NRF52X is enabled (=y).

carlescufi replied:
Agreed with @cvinayak, CONFIG_ARM_MPU_NRF52X sounds best to me as well

fvincenzo replied:
In my opinion this parameter should be considered vendor specific and can be called in any way as far as it makes sense :smile: and respects the dependencies. Said that my preference would be NRF52X_ARM_MPU to maintain the same order used for STM32 and Beetle.

galak reviewed (changes requested):
can you send to arm branch

mike-scott commented:
@galak Woops.. arm branch right.  Would it be ok to finish the initial review here so we can keep these comments and then re-post once @cvinayak and @carlescufi are ok w/ the patch?

galak commented:
Sure no issues w/waiting til comments are agreed or when you repost a version

mike-scott commented:
CHANGES:
- Updated the memory regions slightly to add the ARM M4 Private Peripheral area (which is also included in the STM32 and ARM Beetle configs).
- merged the 2 peripheral areas into 1 so we don't use too many areas
- Fixed compile warning by adding "SOC_FAMILY_NRF5" to the "depends on SOC_FAMILY_ARM || SOC_FAMILY_STM32 || SOC_FAMILY_NRF5" line in the ARM_MPU config.. NOTE: Currently, only nRF52x is supported but I emulated the other checks here for STM32 and ARM where not all of their devices are supported... but probably will be in the future.

mike-scott commented:
Also removed Change-Id.

mike-scott commented:
Rebased on arm branch and addressed @carlescufi  / @cvinayak comments

mike-scott commented on arch/arm/soc/nordic_nrf5/nrf52/mpu_mem_cfg.h:
```diff
@@ -0,0 +1,42 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef _ARM_MPU_MEM_CFG_H_
```
Woops.. need to fix the header guard.

galak commented:
Would like to see @carlescufi / @cvinayak approve this from a review point of view

carlescufi reviewed (approved):
Looks fine to me, nothing to add.

mike-scott commented:
Changes:
- Rebased on latest ARM branch
- Fixed up the mpu_mem_cfg.h header guard: 
#ifndef _NRF52X_MPU_MEM_CFG_H_

galak committed:
Merge branch 'arm' into review-master
